### ISSUE TITLE
Associated constants in traits can not be used in const generics

### ISSUE THREAD

**Issue by newpavlov** [rust-lang/rust#60551]
Initially reported [here](https://github.com/rust-lang/rust/issues/44580#issuecomment-489294181).

The following code:
```rust
trait Foo {
    const N: usize;
    fn foo() -> [u8; Self::N];
}
```
Results in a "no associated item named `N` found for type `Self` in the current scope" compilation error on current Nightly.

Note that associated constants in `impl` blocks [work](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=ecc943d64d35c2cde6f50603d30731c5) without any issues.

cc @varkor @yodaldevoid

**Comment by dtolnay** [comment]
Potentially related to https://github.com/rust-lang/rust/issues/43408.

**Comment by Dylan-DPC** [comment]
The error now is:
```
error: constant expression depends on a generic parameter
 --> src/lib.rs:5:17
  |
5 |     fn foo() -> [u8; Self::N];
  |                 ^^^^^^^^^^^^^
  |
  = note: this may fail depending on what value the parameter takes
```

**Comment by varkor** [comment]
The error with `min_const_generics` enabled is:
```
error: generic parameters must not be used inside of non trivial constant values
 --> src/lib.rs:5:22
  |
5 |     fn foo() -> [u8; Self::N];
  |                      ^^^^^^^ non-trivial anonymous constants must not depend on the parameter `Self`
  |
  = note: type parameters are currently not permitted in anonymous constants
```
which can definitely be improved.

**Comment by varkor** [comment]
I think we ought to prioritise good diagnostics for `min_const_generics` from the beginning, so I'm making this issue as a blocker too.

**Comment by LeSeulArtichaut** [comment]
#77825 landed, should `const-generics-bad-diagnostics` (and `const-generics-blocking`) be removed?

**Comment by varkor** [comment]
Yes, thanks.

**Comment by parraman** [comment]
I think the following problem is related to this issue. If not, I apologize in advance. I don't quite understand why this code works:
```rust
#![feature(min_const_generics)]

struct WithArray<T, const SIZE: usize> {
    data: [T; SIZE]
}
```
while this other one doesn't:
```rust
trait WithAConstant {
    const SIZE: usize;
}

struct WithArray<T, U: WithAConstant> {
    data: [T; U::SIZE]
}
```
This last one produces the following error:
```
error: generic parameters may not be used in const operations
  --> src/main.rs:13:27
   |
13 |     data: [T; U::SIZE]
   |                           ^^^^^^^ cannot perform const operation using `U`
   |
   = note: type parameters may not be used in const expressions
```
Thanks!

**Comment by ChristopherRabotin** [comment]
Hi there,

Is there a fix to this issue on the stable branch of Rust? This issue happens with `rustc 1.52.1 (9bc8c42bb 2021-05-09)`.

[Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=86b2d396c559aa13ae23917741ec36fe) 

Thanks

**Comment by varkor** [comment]
@parraman: sorry, I missed your comment. Your example should eventually work; it is a current implementation limitation. If you enable `#![feature(const_evaluatable_checked)]`, it almost works (and probably can be made to work with a little tweaking to ensure `U`'s variance is known):
```rust
#![feature(const_generics)]
#![feature(const_evaluatable_checked)]

trait WithAConstant {
    const SIZE: usize;
}

struct WithArray<T, U: WithAConstant> where [(); U::SIZE]: {
    data: [T; U::SIZE]
}
```

@ChristopherRabotin: this is a limitation of the current stable version of const generics that we hope to fix in a future version.

**Comment by ChristopherRabotin** [comment]
Okay thanks for the quick answer @varkor . If you were to bet, would you say this will be fix in the the next six months or more? I'm asking because I prefer keeping my crate working with rust stable, but I would be OK switching to nightly for a few months. Thanks

**Comment by varkor** [comment]
There are still a number of design and implementation questions that need to be resolved with `const_evaluatable_checked` before we can consider stabilising it, so likely not in the next six months.

**Comment by hniksic** [comment]
I encountered this limitation, and it makes const generics much harder to apply to my use case, where the behavior of a large chunk of code is fully customized by the trait. Customization normally includes associated types, and should also be able to include associated constants to be used in const generics. For example:

```rust
// n-dimensional point
struct Point<const NDims: usize>([f64; NDims]);

// configuration of various things, including number of dimensions
pub trait Config {
    const NDims: usize;
    // more config here...
}

// use Point in structs generic over C: Config
struct ColorSpace<C: Config> {
    data: Vec<(Point<C::NDims>, u8)>,
}
```

I understand that there are issues to resolve with the more general const evaluation (e.g. panic handling), but couldn't it be possible to refer to trait associated constants as a special case? I.e. have the above compile, but not the more complex `Point<{C::NDims * 2}>` or `Point<{arbitrary_const_fn(C::NDims)}>`?

Currently the only fix is to make the trait generic over `NDims`, but it's both conceptually wrong (the trait needs to choose `NDims` that makes sense for the configuration) and makes the API harder to use because an additional const generic creeps in everywhere. If `Config` needed to define multiple such constants, the problem would be that much worse.

**Comment by josephg** [comment]
Just chiming in to say I'm running into the same problem.

I have a bunch of configuration parameters to tune a custom b-tree. Passing in a trait object for configuration makes my code much cleaner here; if for no other reason so I don't need to copy-paste `<E: EntryTraits, I: TreeIndex<E>, const INTERNAL_CHILDREN: usize, const INTERNAL_CHILDREN: usize>, ...>` in about 20 places throughout my code.

I'm running into the same problem as @hniksic. I'm running into the awkward problem that I can't use the const field members in my trait to size some internal arrays. Now that const generics have landed, I'd love some love on this!

**Comment by newpavlov** [comment]
It looks like the following code successfully [compiles](https://play.rust-lang.org/?version=nightly&gist=0cc7fe072960d110d48e995c9f4e3d8a) on nigthly:
```rust
#![feature(generic_const_exprs)]

trait Foo {
    const N: usize;
    fn foo() -> [u8; Self::N];
}
```

Should I close this issue in favor of #76560?

**Comment by hniksic** [comment]
@newpavlov Please note that the example from the previous comment (that defines `ColorSpace` type) [deosn't compile](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=8906ddca790fc98d00ab35b3a11d08a1) on current nightly. The compiler requires changing `Point<C::NDims>` to `Point<{C::NDims}>` because it treats `C::NDims` as a full-featured expression rather than a constant. But making that change [still fails to compile](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=53d59377c9e681307fdd91606d00a966), citing an "unconstrained generic constant" and suggesting a bound of `where [(); {C::NDims}]`. Adding that bound leads to [further errors](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=5a7870fa41ef3ef7b80da493e3d0a749) about "unused parameter `C`" which I'm not sure how to address.

It would be nice to have this resolved even without the full support for expressions in generic parameters. But even if that's not possible, the feature doesn't appear to work just yet even with `generic_const_exprs`, so the issue shouldn't be closed at this point.

**Comment by tarcieri** [comment]
Somewhat related to this issue that I just discovered today: as an alternative to constants, it's possible to parameterize a trait with a const generic today on stable Rust using only `min_const_generics`:

https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=1b2ad1c2f09b2c468f78b393a42e1dad

```rust
trait Foo<const N: usize> {
    fn do_x(&self) -> [u8; N];
}

struct Bar;

impl Foo<42> for Bar {
    fn do_x(&self) -> [u8; 42] {
        [0u8; 42]
    }
}
```

It's not quite the same thing, but it might be good enough for many of these potential applications.

**Comment by mooman219** [comment]
I'm trying to write this on nightly with `#![feature(generic_const_exprs)]`

```rust
pub trait Foo {
    const COUNT: usize;
    const NAMES: [&'static str; Self::COUNT];
}

pub struct Bar<T: Foo> {
    name_lookup: [resource::UniformLocation; T::COUNT]
}
```
But I cannot 
```
error: unconstrained generic constant
  --> src\test.rs:18:5
   |
18 |     const NAMES: [&'static str; Self::COUNT];
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: try adding a `where` bound using this expression: `where [(); Self::COUNT]:`

error: unconstrained generic constant
  --> src\test.rs:22:18
   |
22 |     name_lookup: [resource::UniformLocation; T::COUNT],
   |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: try adding a `where` bound using this expression: `where [(); T::COUNT]:`
```
I'm trying to avoid writing `pub trait Foo<const COUNT: usize>` because it results in a lot of boilerplate I need to propagate : (.

**Comment by linclelinkpart5** [comment]
> I'm trying to write this on nightly with `#![feature(generic_const_exprs)]`
> 
> ```rust
> pub trait Foo {
>     const COUNT: usize;
>     const NAMES: [&'static str; Self::COUNT];
> }
> 
> pub struct Bar<T: Foo> {
>     name_lookup: [resource::UniformLocation; T::COUNT]
> }
> ```
> 
> But I cannot
> 
> ```
> error: unconstrained generic constant
>   --> src\test.rs:18:5
>    |
> 18 |     const NAMES: [&'static str; Self::COUNT];
>    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>    |
>    = help: try adding a `where` bound using this expression: `where [(); Self::COUNT]:`
> 
> error: unconstrained generic constant
>   --> src\test.rs:22:18
>    |
> 22 |     name_lookup: [resource::UniformLocation; T::COUNT],
>    |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>    |
>    = help: try adding a `where` bound using this expression: `where [(); T::COUNT]:`
> ```
> 
> I'm trying to avoid writing `pub trait Foo<const COUNT: usize>` because it results in a lot of boilerplate I need to propagate : (.

I'm in this exact same scenario, I'm trying to avoid propagating the `N` generic parameter everywhere in my codebase.

**Comment by porky11** [comment]
I have a similar problem:

```rust
trait VectorSpace
where
    Self: ...,
    Self: Into<[Self::Scalar, Self::D]> + From<[Self::Scalar, Self::D]>,
{
    type Scalar;
    const D: usize;
}
```
Wouldn't this be the next thing, that should be stabilized? Doesn't seem too complicated.

**Comment by rcarson3** [comment]
I believe this is related to this issue as well on nightly, but I'm running into the following issue with a more complicated piece of code I'm running on. However, I've worked out a simpler example here:
https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=f15f265d88a25cfa9fc8822cf3c5950e

The compiler can't figure out that the bound is already met even though all the values for it are known at compile time. I've run into similar issues with this in the past where I have arrays with defined const size but the compiler can't have them swap places, and it always wants me to add additional where clauses for those things...

```Rust
#![no_std]
#![allow(incomplete_features)]
#![feature(generic_const_exprs)]

trait MathTrait {

    const NDIM: usize;
    fn calculate_math(&mut self, data: &mut [f64]);
    
}

struct Solver<'a, M>
where
    M: MathTrait + Sized,
    [f64; M::NDIM]: Sized,
{
    pub x: [f64; M::NDIM],
    problem: &'a mut M,
}

impl<'a, M> Solver<'a, M> 
where
    M: MathTrait + Sized,
    [f64; M::NDIM]: Sized,
{
    pub fn new(problem: &'a mut M ) -> Solver<'a, M> {
        Solver::<'a, M> {
            x: [0.0_f64; M::NDIM],
            problem,
        }
    }
}

trait PL {
    const NSIZE: usize;
}

struct PhysicsComp{}

impl PL for PhysicsComp {
    const NSIZE: usize = 32;
}

struct Physics<P: PL> {
    comp: P,
}

impl<P: PL> MathTrait for Physics<P> 
where
    [f64; P::NSIZE]: Sized,
{
    const NDIM: usize = P::NSIZE;
    
    fn calculate_math(&mut self, data: &mut [f64]) {
        /* do math here or whatever */
    }
}

struct DoStuff<P: PL> {
    data: P,
}

impl<P: PL> DoStuff<P>
where
    [f64; P::NSIZE]: Sized, // Bound that is equivalent to the error of [f64; M::NDIM]: Sized,
{
    fn problem(&self) -> bool {
        let mut phys = Physics {
            comp: PhysicsComp {},  
        };
        // This line fails quite hard
        let mut solver = Solver::<Physics::<P>>::new(&mut phys);
        true
    }
}


fn main() {
    let failure = DoStuff {
        data: PhysicsComp {},
    };
    failure.problem();
}
```

```
error[[E0599]](https://doc.rust-lang.org/nightly/error-index.html#E0599): no function or associated item named `new` found for struct `Solver<'_, Physics<P>>` in the current scope
  [--> src/main.rs:76:50
](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=f15f265d88a25cfa9fc8822cf3c5950e#)   |
12 | / struct Solver<'a, M>
13 | | where
14 | |     M: MathTrait + Sized,
15 | |     [f64; M::NDIM]: Sized,
...  |
18 | |     problem: &'a mut M,
19 | | }
   | |_- function or associated item `new` not found for this
...
76 |           let mut solver = Solver::<Physics::<P>>::new(&mut phys);
   |                                                    ^^^ function or associated item cannot be called on `Solver<'_, Physics<P>>` due to unsatisfied trait bounds

error: unconstrained generic constant
  [--> src/main.rs:76:26
](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=f15f265d88a25cfa9fc8822cf3c5950e#)   |
76 |         let mut solver = Solver::<Physics::<P>>::new(&mut phys);
   |                          ^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: try adding a `where` bound using this expression: `where [(); M::NDIM]:`
note: required by a bound in `Solver`
  [--> src/main.rs:15:11
](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=f15f265d88a25cfa9fc8822cf3c5950e#)   |
12 | struct Solver<'a, M>
   |        ------ required by a bound in this
...
15 |     [f64; M::NDIM]: Sized,
   |           ^^^^^^^ required by this bound in `Solver`

For more information about this error, try `rustc --explain E0599`.
```

**Comment by burdges** [comment]
If I understand, rustc cannot correctly identify some bound when using the associated constant internally to a default method?  https://play.rust-lang.org/?version=nightly&mode=debug&edition=2015&gist=e1114345b12197fdb922ef1acb2e80ae

**Comment by chrismooredev** [comment]
Similarly to burdges' playground, I'd also like to throw in the ugly syntax necessary for using bool associated constants. It requiring that bound, and especially having to cast it to a usize, is not intuitive.

I planned to use this feature similarly to my playground, have a trait with a ton of default implementations for a few types, while still being able to use const type parameters as feature/optimization flags for different internal algorithms, operating on those types.

[Playground Link](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=3f2199972a25b9926c67d5e5b0a22e16)
```rs
#![feature(generic_const_exprs)]

fn _internal_impl<const FLAG: bool>() -> &'static str {
    if FLAG {
        "run for accuracy"
    } else {
        "run for speed"
    }
}

trait Operation {
    const FLAG: bool;
    fn run() -> &'static str where [(); Self::FLAG as usize]: {
        // default implementation
        _internal_impl::<{ Self::FLAG }>()
    }
}
```

**Comment by alexpyattaev** [comment]
Can someone explain to a mere mortal what is the purpose of 
```rust
where [(); TRAIT::CONST_NAME] :
```
syntax? What does it achieve in particular? I've seen something about variance but it gets muddy from there.

**Comment by tgross35** [comment]
> Can someone explain to a mere mortal what is the purpose of
> 
> ```rust
> where [(); AssociatedType::CONST_NAME] :
> ```

My very naive understanding is that without the bound either (1) `[T; TypeA::LEN]` and `[T; TypeB::LEN]` would be considered the same in the type system even if `TypeA::LEN != TypeB::LEN` (because associated consts behave differently from associated types) or (2) a `[T; TypeA::LEN]` in two different places would not evaluate to the same type because it has no covariance to itself. I'm not sure which of these two are more accurate, but I believe that it's a result of `[T; N]` being covariant to `T` but having no relationship to `N` - and I think that is what this syntax is constraining.

I don't think the syntax is intended to stay - at least I sure hope not, it's pretty unfriendly and doesn't at all hint what it's doing. I think it's just some existing syntax that happens to do the trick. If it _needs_ to be expressed, then some `std::ops` type traits would imho be a better choice:

```rust
struct Foo<T, U: WithAConstant>
    // Only one of these options
    where AnyArray<U::Size>: Covariant
    where AnyArray<U::Size>: CovariantOn<U::SIZE>
{
    data: [T; U::SIZE]
}
```

But I think the compiler should be able to infer the correct usage and just error if something is too complex for a simple representation. At least I hope that's possible, because any sort of bound is a lot of noise for something that "should work" in the default case.

If inference isn't possible, I'd be kind of curious to see some examples explaining why. Or at least a better explanation from somebody about what the actual variance issue is

**Comment by alexpyattaev** [comment]
Well this makes the purpose clear. As you suggested, this will likely go away eventually, since for more "normal" cases the conversion from [T; N] to [T;3] when N=3 happens automatically without messy syntax.

The bigger question that is still bugging me is how someone came up with that construct and what does it truly mean...

**Comment by tgross35** [comment]
It doesn't seem like anybody came up with the syntax specifically for this feature, I just think it's something that exists in the type system already. For example, on stable this compiles (even though it doesn't do anything):

```rust
struct Foo where [(); 100]: {}
```

And it seems like you can use any type, `()` just sort of hints "can be any type" even though that's not exactly what it says. This compiles on nightly:

```rust
#![feature(generic_const_exprs)]

trait WithAConstant {
    const SIZE: usize;
}

struct WithArray<T, U: WithAConstant>
where [String; U::SIZE]: // using any type here has the same result
{
    data: [T; U::SIZE]
}
```

Since the syntax already exists, I don't think anything syntax-related would at all block stabilization of `generic_const_exprs`. I think it makes sense instead to create a separate feature like `const_variance_inference` that has the purpose of eliminating its need, which could stabilize on its own time.

But I don't know if this has been discussed anywhere already, and would be curious to hear from anybody more involved in the RFC / decision making of this feature.

As far as status of stabilizing `generic_const_exprs` It seems like there are quite a few issues open that be blocking https://github.com/rust-lang/project-const-generics/issues?q=is%3Aissue+is%3Aopen+label%3AA-generic-exprs

(Not sure if rustbot will let me do this since it's not my issue, but seems like this one belongs in that group too since it's related)

@rustbot label +A-generic-exprs

**Comment by hniksic** [comment]
@igorechek06 It's great that you found a workaround that works for you, but I need to point out that your solution doesn't help when you need the trait _implementation_ to specify the value of the constant (`SIZE` in your example). That's the situation in the description of this issue, as well as the one presented in [my earlier comment](https://github.com/rust-lang/rust/issues/60551#issuecomment-882031857).
